homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,576 @@
+            self._state = float(temp_state)
+            self._unit = TEMP_FAHRENHEIT
+
+        return self._state
+
+
+class OmniLogicPumpSpeedSensor(OmnilogicSensor):
+    """Define an OmniLogic Pump Speed Sensor."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        backyard: dict,
+        bow: dict,
+        device_class: str,
+        icon: str,
+        unit: str,
+        sensor_data: dict,
+    ):
+        """Initialize the sensor."""
+
+        name = f"{backyard['BackyardName']} {bow['Name']} {name}"
+
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            device_class=device_class,
+            icon=icon,
+            unit=unit,
+            sensor_data=sensor_data,
+            system_id=sensor_data.get("systemId"),
+        )
+
+    @property
+    def state(self):
+        """Return the state for the pump speed sensor."""
+
+        sensor_data = {}
+        pump_type = ""
+        pump_speed = 0
+
+        if self._kind == "filter_pump_speed":
+            for backyard in self.coordinator.data:
+                for bow in backyard.get("BOWS"):
+                    if bow["Filter"]["systemId"] == self._system_id:
+                        sensor_data = bow["Filter"]
+                        break
+
+            pump_type = PUMP_TYPES.get(sensor_data.get("Filter-Type"))
+            pump_speed = sensor_data.get("filterSpeed")
+
+        elif self._kind == "pump_speed":
+            for backyard in self.coordinator.data:
+                for bow in backyard.get("BOWS"):
+                    for pump in bow.get("Pumps"):
+                        if pump["systemId"] == self.entity_data["systemId"]:
+                            sensor_data = pump
+                            break
+
+            if sensor_data.get("Filter-Type") == "PMP_VARIABLE_SPEED_PUMP":
+                pump_type = "VARIABLE"
+            elif sensor_data.get("Filter-Type") == "PMP_SINGLE_SPEED":
+                pump_type = "SINGLE"
+            elif sensor_data.get("Filter-Type") == "PMP_DUAL_SPEED":
+                pump_type = "DUAL"
+
+            pump_speed = sensor_data.get("pumpSpeed")
+
+        if pump_type == "VARIABLE":
+            self._unit = "%"
+            self._state = pump_speed
+        elif pump_type == "DUAL":
+            if pump_speed == 0:
+                self._state = "off"
+            elif pump_speed == sensor_data.get("Min-Pump-Speed"):
+                self._state = "low"
+            elif pump_speed == sensor_data.get("Max-Pump-Speed"):
+                self._state = "high"
+        elif pump_type == "SINGLE":
+            if pump_speed == 0:
+                self.state = "off"
+            elif pump_speed == sensor_data.get("Max-Pump-Speed"):
+                self._state = "on"
+
+        self._attrs["PumpType"] = pump_type
```

Comment:
State attribute keys must be lowercase snake_case.
```suggestion
        self._attrs["pump_type"] = pump_type
```

Reply:
Addressed